src/bin/ion/commands/jq.rs

```diff
@@ -173,11 +202,148 @@ impl PartialOrd for JaqElement {
+    fn to_big_decimal(self) -> BigDecimal {
+        self
+    }
+
+    fn to_decimal(self) -> Decimal {
+        let (coeff, exponent) = self.into_bigint_and_exponent();
+        let data = coeff.to_i128().unwrap();
+        Decimal::new(data, -exponent)
+    }
+}
+
+trait FloatMath {
+    fn to_f64(self) -> Option<f64>;
+}
+
+impl FloatMath for f64 {
+    fn to_f64(self) -> Option<f64> {
+        Some(self)
+    }
+}
+
+impl FloatMath for Int {
+    fn to_f64(self) -> Option<f64> {
+        self.as_i128().and_then(|data| {
+            let float = data as f64;
+            (float as i128 == data).then_some(float)
+        })
+    }
+}
+
+impl FloatMath for Decimal {
+    fn to_f64(self) -> Option<f64> {
+        self.to_big_decimal().to_f64()
+    }
+}
+
+impl FloatMath for Value {
+    fn to_f64(self) -> Option<f64> {
+        match self {
+            Value::Int(i) => i.to_f64(),
+            Value::Decimal(d) => d.to_f64(),
+            _ => None
+        }
+    }
+}
+
 impl Add for JaqElement {
     type Output = ValR<Self>;
 
     fn add(self, _rhs: Self) -> Self::Output {
-        todo!()
+        let (lhv, rhv) = (self.into_value(), _rhs.into_value());
+
+        use Value::*;
+
+        let elt: Element = match (lhv, rhv) {
+            (List(lhs), List(rhs)) => {
+                ion_rs::List::from_iter(lhs.into_iter().chain(rhs)).into()
+            },
+            (SExp(lhs), SExp(rhs)) => {
+                ion_rs::SExp::from_iter(lhs.into_iter().chain(rhs)).into()
+            },
+            (String(lhs), String(rhs)) => {
+                format!("{}{}", lhs.text(), rhs.text()).into()
+            }
+            (Struct(lhs), Struct(rhs)) => {
+                //TODO: Recursively remove duplicate fields, first field position but rhs and last field wins
+                lhs.clone_builder()
+                    .with_fields(rhs.fields())
+                    .build().into()
+            }
+
+            // Number types, only lossless operations
+            (Int(lhs), Int(rhs)) => (lhs + rhs).into(),
+            (Float(lhs), Float(rhs)) => (lhs + rhs).into(),
+            (Decimal(lhs), Decimal(rhs)) => lhs.add(rhs).into(),
+            (Decimal(dv), Int(iv)) | (Int(iv), Decimal(dv)) => dv.add(iv).into(),
+
+            // jq treats JSON's untyped null as an additive identity, e.g. 0 / "" / [] / {}
+            (lhs, Null(IonType::Null)) => lhs.into(),
+            (Null(IonType::Null), rhs) => rhs.into(),
+
+            // Typed nulls we must handle differently, we can only add similar types
+            (Null(lht), Null(rht)) if lht == rht => Null(lht).into(),
+            (Null(lht), rhs) if lht == rhs.ion_type() => rhs.into(),
+            (lhs, Null(rht)) if lhs.ion_type() == rht => lhs.into(),
+
+            // Only try potentially lossy Float conversions when we've run out of the other options
+            (v, Float(fv)) | (Float(fv), v) if matches!(v.ion_type(), IonType::Int | IonType::Decimal) => {
```

Comment:
What are the pros/cons to lowering the Decimal to float, rather than promoting the float to decimal? I'm just curious, I'd assume promoting the float would be less lossy.